ambry-store/src/main/java/com.github.ambry.store/BlobStore.java

```diff
@@ -167,27 +166,27 @@ public void put(MessageWriteSet messageSetToWrite) throws StoreException {
       synchronized (lock) {
         // Validate that log end offset was not changed. If changed, check once again for existing
         // keys in store
-        long currentIndexEndOffset = index.getCurrentEndOffset();
-        if (currentIndexEndOffset != indexEndOffsetBeforeCheck) {
+        Offset currentIndexEndOffset = index.getCurrentEndOffset();
+        if (!currentIndexEndOffset.equals(indexEndOffsetBeforeCheck)) {
           FileSpan fileSpan = new FileSpan(indexEndOffsetBeforeCheck, currentIndexEndOffset);
           for (MessageInfo info : messageSetToWrite.getMessageSetInfo()) {
             if (index.findKey(info.getStoreKey(), fileSpan) != null) {
               throw new StoreException("Key already exists on filespan check", StoreErrorCodes.Already_Exist);
             }
           }
         }
-        // TODO (Index changes): Working under the assumption that there is only one log segment.
-        long currentLogEndOffset = log.getEndOffset().getOffset();
+        Offset currentLogEndOffset = log.getEndOffset();
         messageSetToWrite.writeTo(log);
```

Comment:
This makes me a little uncomfortable. I feel that the `IndexValue`s should be created in a more tightly coupled way. There is a bit of an abstraction leakage here in that the `Log` determines where the individual messages go and then the `Store` sort of "figures" out the offsets of these messages in the log in creating the `IndexValue`s - and does not even consult the `Log` for that.

I understand that the old code did it this way as well - but at least in the old code it was a simple addition. Now that the offset additions are more sophisticated, for one I think we should let just one class deal with that. Secondly, since we made the change to write each message in the `writeSet` individually to the `Log` - perhaps the `IndexValue`s can be created after each individual message? Perhaps the `writeSet` could expose an iterator-like interface:
```java
while (writeSet.hasNext()) {
  writeSet.writeNext(); // or some such thing.
  ... generate the `Offset` of the last message with help from the Log.
}
```